Mathlib/RingTheory/MvPowerSeries/LinearTopology.lean

```diff
@@ -0,0 +1,203 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+import Mathlib.Data.Finsupp.Interval
+import Mathlib.RingTheory.MvPowerSeries.PiTopology
+import Mathlib.Topology.Algebra.LinearTopology
+import Mathlib.Topology.Algebra.Nonarchimedean.Bases
+import Mathlib.RingTheory.TwoSidedIdeal.Operations
+
+/-! # Linear topology on the ring of multivariate power series
+
+- `MvPowerSeries.basis`: the ideals of the ring of multivariate power series
+all coefficients the exponent of which is smaller than some bound vanish.
+
+## Instances :
+
+- `MvPowerSeries.isLinearTopology` : if `R` has a linear topology,
+then the product topology on `MvPowerSeries σ R` is a linear topology.
+This applies in particular when `R` has a discrete topology.
+
+-/
+
+namespace MvPowerSeries
+
+namespace LinearTopology
+
+open scoped Topology
+
+open Set SetLike
+
+/-- The underlying family for the basis of ideals in a multivariate power series ring. -/
+def basis (σ : Type*) (R : Type*) [Ring R] (Jd : TwoSidedIdeal R × (σ →₀ ℕ)) :
+    TwoSidedIdeal (MvPowerSeries σ R) := by
+  apply TwoSidedIdeal.mk' {f | ∀ e ≤ Jd.2, coeff R e f ∈ Jd.1}
+  · simp [coeff_zero]
+  · exact fun hf hg e he ↦ by rw [map_add]; exact add_mem (hf e he) (hg e he)
+  · exact fun {f} hf e he ↦ by simp only [map_neg, neg_mem, hf e he]
+  · exact fun {f g} hg e he ↦ by
+      classical
+      rw [coeff_mul]
+      apply sum_mem
+      rintro uv huv
+      apply TwoSidedIdeal.mul_mem_left
+      exact hg _ (le_trans (le_iff_exists_add'.mpr
+        ⟨uv.fst, (Finset.mem_antidiagonal.mp huv).symm⟩) he)
+  · exact fun {f g} hf e he ↦ by
+      classical
+      rw [coeff_mul]
+      apply sum_mem
+      rintro uv huv
+      apply TwoSidedIdeal.mul_mem_right
+      exact hf _ (le_trans (le_iff_exists_add.mpr ⟨uv.2, (Finset.mem_antidiagonal.mp huv).symm⟩) he)
+
+variable {σ : Type*} {R : Type*} [Ring R]
+
+/-- A power series `f` belongs to the twosided ideal `basis σ R ⟨J, d⟩`
+if and only if `coeff R e f ∈ J` for all `e ≤ d`. -/
+theorem mem_basis_iff {f : MvPowerSeries σ R} {Jd : TwoSidedIdeal R × (σ →₀ ℕ)} :
+    f ∈ basis σ R Jd ↔ ∀ e ≤ Jd.2, coeff R e f ∈ Jd.1 := by
+  simp [basis]
+
+/-- If `J ≤ K` and `e ≤ d`, then we have the inclusion of twosided ideals
+`basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e,>`. -/
+theorem basis_le {Jd Ke : TwoSidedIdeal R × (σ →₀ ℕ)} (hJK : Jd.1 ≤ Ke.1) (hed : Ke.2 ≤ Jd.2) :
+    basis σ R Jd ≤ basis σ R Ke :=
+  fun _ ↦ forall_imp (fun _ h hue ↦ hJK (h (le_trans hue hed)))
+
+/-- `basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e⟩` if and only if `J ≤ K` and `e ≤ d`. -/
+theorem basis_le_iff {J K : TwoSidedIdeal R} {d e : σ →₀ ℕ} (hK : K ≠ ⊤) :
+    basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e⟩ ↔ J ≤ K ∧ e ≤ d := by
```

Comment:
What is the use case for this theorem? It seems to me that we would always use the easy direction `basis_le` instead, but I may be missing something.

Reply:
Unclear.